core/src/main/java/org/pentaho/di/osgi/KarafLifecycleListener.java

```diff
@@ -129,66 +142,119 @@ private void startTimeoutThread() {
+    ServiceReference<T> serviceReference = bundleContext.getServiceReference( serviceClass );
+    if ( serviceReference == null ) {
+      return null;
+    }
+    return bundleContext.getService( serviceReference );
+  }
+
+  /**
+   * Actively wait until features are installed.
+   *
+   * The current used implementation of IKarafFeatureWatcher, which is org.pentaho.osgi.impl.KarafFeatureWatcherImpl,
+   * is only properly functioning with booting Karaf with empty caches. This is because of a behaviour (bug?) of the
+   * Karaf Feature Service that when booting with full caches, it hydrates the persisted state where all features have
+   * been installed and started. As such, when asking the Feature Service if a given feature is installed, it will
+   * reply positively even if its bundles haven't been started yet.
+   */
+  private void waitForFeatures() {
+    IKarafFeatureWatcher karafFeatureWatcher = getOsgiService( IKarafFeatureWatcher.class );
+    try {
+      if ( karafFeatureWatcher == null ) {
+        throw new IKarafFeatureWatcher.FeatureWatcherException( "No IKarafFeatureWatcher service available." );
+      }
+      karafFeatureWatcher.waitForFeatures();
+    } catch ( IKarafFeatureWatcher.FeatureWatcherException e ) {
+      logger.error( "Error in Feature Watcher", e );
+    }
+  }
+
+  private void waitForBlueprints() {
+    IKarafBlueprintWatcher karafBlueprintWatcher = getOsgiService( IKarafBlueprintWatcher.class );
+    try {
+      if ( karafBlueprintWatcher == null ) {
+        throw new IKarafBlueprintWatcher.BlueprintWatcherException( "No IKarafBlueprintWatcher service available." );
+      }
+      karafBlueprintWatcher.waitForBlueprint();
+    } catch ( IKarafBlueprintWatcher.BlueprintWatcherException e ) {
+      logger.error( "Error in Blueprint Watcher", e );
+    }
+  }
+
+  private void acceptEventOnDelayedServiceNotifiersDone() {
+    final AtomicBoolean accepted = new AtomicBoolean( false );
+    DelayedServiceNotifierListener delayedServiceNotifierListener = new DelayedServiceNotifierListener() {
+      @Override public void onRun( LifecycleEvent lifecycleEvent, Object serviceObject ) {
+        if ( osgiPluginTracker.getOutstandingServiceNotifierListeners() == 0 && !accepted.getAndSet( true ) ) {
+          logger.debug( "Done waiting on delayed service notifiers" );
+          event.accept();
+          osgiPluginTracker.removeDelayedServiceNotifierListener( this );
+        }
+      }
+    };
+
+    logger.debug( "About to start waiting on delayed service notifiers" );
+    osgiPluginTracker.addDelayedServiceNotifierListener( delayedServiceNotifierListener );
+    delayedServiceNotifierListener.onRun( null, null );
+
+  }
+
+  /**
+   * Actively wait until the OSGi framework has started.
+   * With Karaf empty caches this means wait until all bundles defined in the startup.properties have started.
+   * With Karaf full caches this means wait until all cached bundles have started.
+   */
```

Comment:
What would really be the excepted behaviour? To wait for all bundles or only the startup bundles? Is this probable to change in future versions?

Reply:
IMO the expected behavior should be to only mark the framework as started when both all startup.properties bundles and boot features bundles have started. As such, the expected behaviour IMO is the one we get with hot caches. 
However, because of the "non-linear" resolving of packages and hence bundles by the feature resolver, knowing / precalculating all of these bundles with cold caches is probably not easy.

The difference in behaviour between cold and hot caches is defined by karaf in the bundles that are included in the framework start. I believe this is unlikely to change much in future versions.